pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,354 @@
+	// In this call, it is possible to get errors back with non-nil upstreamSpec.
+	// If upstreamSpec is nil then the syncing failed for some reason. This is reported to the user, and this function returns at the end of this if-statement.
+	// If upstreamSpec is non-nil then the syncing occurred as expected, but the node groups have health issues that are reported to the user.
+	// In this second case, the message is set on the Updated condition, but execution continues because the sync was successful.
+	upstreamConfig, err := GetComparableUpstreamSpec(c.secretsCache, cluster)
+	if err != nil {
+		var statusErr error
+		var syncFailed string
+		if upstreamConfig == nil {
+			syncFailed = ": syncing failed"
+		}
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.False(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, fmt.Sprintf("[Syncing error%s] %s", syncFailed, err.Error()))
+
+		cluster, statusErr = c.clusterClient.Update(cluster)
+		if statusErr != nil {
+			return cluster, statusErr
+		}
+	} else if strings.Contains(apimgmtv3.ClusterConditionUpdated.GetMessage(cluster), "[Syncing error") {
+		cluster = cluster.DeepCopy()
+		apimgmtv3.ClusterConditionUpdated.True(cluster)
+		apimgmtv3.ClusterConditionUpdated.Message(cluster, "")
+		cluster, err = c.clusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+	if upstreamConfig == nil {
+		return cluster, err
+	}
+
+	switch {
+	case cluster.Spec.AKSConfig != nil:
+		if !reflect.DeepEqual(cluster.Status.AKSStatus.UpstreamSpec, upstreamConfig.aksConfig) {
+			logrus.Infof("updating cluster [%s], upstream change detected", cluster.Name)
+			cluster = cluster.DeepCopy()
+			cluster.Status.AKSStatus.UpstreamSpec = upstreamConfig.aksConfig
+			cluster, err = c.clusterClient.Update(cluster)
+			if err != nil {
+				return cluster, err
+			}
+		}
+
+		if !reflect.DeepEqual(cluster.Spec.AKSConfig, cluster.Status.AKSStatus.UpstreamSpec) {
+			logrus.Infof("cluster [%s] currently updating, skipping spec sync", cluster.Name)
+			return cluster, nil
+		}
+
+		// check for changes between EKS spec on cluster and the EKS spec on the EKSClusterConfig object
+		specMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(cluster.Spec.AKSConfig)
+		if err != nil {
+			return cluster, err
+		}
+
+		upstreamSpecMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(upstreamConfig.aksConfig)
+		if err != nil {
+			return cluster, err
+		}
+
+		var updateConfig bool
+		for key, value := range upstreamSpecMap {
+			if specMap[key] == nil {
+				continue
+			}
+			if reflect.DeepEqual(specMap[key], value) {
+				continue
+			}
+			updateConfig = true
+			specMap[key] = value
+		}
+
+		if !updateConfig {
+			logrus.Infof("cluster [%s] matches upstream, skipping spec sync", cluster.Name)
+			return cluster, nil
+		}
+
+		if err = runtime.DefaultUnstructuredConverter.FromUnstructured(specMap, cluster.Spec.AKSConfig); err != nil {
+			return cluster, err
```

Comment:
This code is the same in the EKS case, and `ToUnstructured` and `FromUnstructured` take `interface{}` as parameters. Doing something like:
```golang
var clusterCloudConfig, upstreamCloudConfig, appliedCloudConfig interface{}
switch {
	case cluster.Spec.AKSConfig != nil:
		  ...
		  clusterCloudConfig = cluster.Spec.AKSConfig
		  appliedCloudConfig = cluster.Status.AppliedSpec.AKSConfig
		  upstreamCloudConfig = upstreamConfig.aksConfig
	
	case cluster.Spec.EKSConfig != nil:
		...
}
if !reflect.DeepEqual(clusterCloudConfig, appliedCloudConfig) {
...
```
If that doesn't work, then put this section of code in a function that takes in these three `interface{}`s and call it in each case of the switch.

Reply:
Thank you for tip, it should be better now